src/vtarget-out/inv-syn/inv_syn.cc

```diff
@@ -0,0 +1,633 @@
+      (not (|%w%_n rst|     |__SwI__|))  ; why not removed?
+      (not (|%w%_n rst|     |__Swst__|)) ; why not removed?
+      (|%w%_h| |__SwBI__|)
+      (|%w%_h| |__SwI__|)
+      (|%w%_h| |__Swst__|)
+      (|%w%_t|              |__SwBI__| |__SwI__|)
+      (|__AMC__wrapper|     |__SwI__|)
+      (|%w%_t|              |__SwI__|  |__Swst__|)
+      (|__AMC__wrapper|     |__Swst__|)
+      (|__ASPT__|           |__Swst__|)
+      (INV1 (|%w%_h %subi%| |__Swst__|))
+    )
+    (INV2 |__Swst__|)
+  )
+)
+
+; inv2 /\ T => inv2
+(rule 
+  (=>
+    (and
+        (INV2 |__Sw__|)
+        (|%w%_t| |__Sw__| |__Sw'__|)
+        (|__AMC__wrapper| |__Sw__|)
+        (|__AMC__wrapper| |__Sw'__|)
+        (not (|%w%_n rst|     |__Sw__|))  ; why not removed?
+        (not (|%w%_n rst|     |__Sw'__|)) ; why not removed?
+        (|%w%_h| |__Sw__|)
+        (|%w%_h| |__Sw'__|)
+    )
+    (INV2 |__Sw'__|)
+  )
+)
+
+
+; inv2 /\ ~ assert => fail
+(rule 
+  (=> 
+    (and
+      (INV2             |__Sw__|)
+      (|__AMC__wrapper| |__Sw__|)
+      (|%w%_h| |__Sw__|)
+      (not (|__ASST__|  |__Sw__|))
+    )
+    fail)
+)
+
+(query fail :print-certificate true)
+
+)***";
+
+// a local helper function to produce a list of and
+std::string JoinListWithAnd(const std::vector<std::string> & l) {
+  if (l.size() == 0)
+    return "true";
+  else if(l.size() == 1)
+    return l[0];
+  return "(and " + Join(l, " ")+")";
+}
+
+size_t skipArgs(const std::string & stmt, size_t start) {
+  // need to fit 
+  auto pos = start;
+  ILA_ASSERT(stmt.at(pos) == '(') << "Expecting '('";
+  unsigned level = 0;
+  do {
+    if (stmt.at(pos) == '(')
+      level ++;
+    else if (stmt.at(pos) == ')')
+      level --;
+    pos ++;
+  } while(level != 0);
+  return pos;
+}
+
+// a local helper function to return the type of a smt (define-fun ) stmt 
+bool is_returntype_bool(const std::string & stmt) {
+  std::string beginning("(define-fun |");
+  ILA_ASSERT (stmt.find(beginning) == 0) << "Not a `(define-fun |` stmt";
+  auto pos =  beginning.length()+1;
+  auto end_func_name_pos = stmt.find('|',pos);
+  auto start_arg_list = stmt.find('(', end_func_name_pos);
+  auto end_arg_list = skipArgs(stmt, start_arg_list);
+  auto Bool_pos = stmt.find("Bool", end_arg_list);
+  auto Bitvec_pos = stmt.find("(_ BitVec", end_arg_list);
+  if (Bool_pos < Bitvec_pos)
+    return true;
+  return false;
+}
+
+
+// extract 
+bool extractSigDefFromLine(
+  const std::string & mod_name,
+  const std::string & line,
+  std::set<std::string> & s, 
+  std::vector<std::pair<std::string,bool>> & v) {
+  // list of (name, is bool?)
+  bool found_match_state = false;
+  std::string search_target_wire_num = "(define-fun |" + mod_name + "#";
+  std::string search_target_n = "(define-fun |" + mod_name + "_n ";
+  if (line.find(search_target_wire_num) == 0) { // begins with it
+    auto mark = line.find("; \\");
+    if (mark != line.npos) {
+      auto signame = line.substr(mark+strlen("; \\"));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)